[Merged by Bors] - simplify calculate_committee_fraction

eserilev opened the pull request:
## Issue Addressed

[#4211](https://github.com/sigp/lighthouse/issues/4211)

## Proposed Changes

This PR conforms the helper function `calculate_committee_fraction` to the [v1.3.0 spec](https://github.com/ethereum/consensus-specs/blob/v1.3.0/specs/phase0/fork-choice.md#get_weight)

## Additional Info

the old definition of `calculate_committee_fraction` is almost identical, but the new definition is simpler.

eserilev committed:
simplify calculate_committee_fraction

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -1055,13 +1055,7 @@ pub fn calculate_committee_fraction<E: EthSpec>(
-        .num_active_validators
-        .checked_div(E::slots_per_epoch())?;
-    let committee_weight = committee_size.checked_mul(average_balance)?;
+    let committee_weight = justified_balances.total_effective_balance;
```
this is missing the division by slots per epoch, isn't it?

eserilev replied:
correct, thanks for catching that

eserilev committed:
divide by slots_per_epoch

eserilev commented:
fork_choice_test_definition::execution_status::test_03 is failing

line 1003 mentions a "magic number" generated from `calculate_committee_fraction`. I might need to generate a new magic number here for the test to pass

michaelsproul commented:
> line 1003 mentions a "magic number" generated from calculate_committee_fraction

Huh, that is strange. I wouldn't have thought the values calculated by `calculate_committee_fraction` would actually be any different due to the simplification. We may want to dig into this a bit to understand exactly what's going on

eserilev committed:
updated magic numbers to allow test to pass

eserilev commented:
> > line 1003 mentions a "magic number" generated from calculate_committee_fraction
> 
> Huh, that is strange. I wouldn't have thought the values calculated by `calculate_committee_fraction` would actually be any different due to the simplification. We may want to dig into this a bit to understand exactly what's going on

## Digging into whats going on

the update to `calculate_committee_fraction` is causing issues with test_03. I attempted to debug the issue by running test_03 locally. When executing the operations in test_03, I added a call to  `calculate_committee_fraction` each time a new op is iterated through, and logged the results.

In the previous version of `calculate_committee_fraction`, calling the function in test_03 returns 31,000 at each op iteration

In the new version of `calculate_committee_fraction`, calling the function in test_03 returns 31,250 at each op iteration.

so the previous vs new iterations return a slightly different value.

 Furthermore, in test_03, the magic numbers chosen on lines 990, 994, and 1003 only work for the older version of `calculate_committee_fraction`. In my most recent push I updated the magic numbers to reflect what they should be for the new version of `calculate_committee_fraction`. This update allows test_03 to pass. To be honest, I have no idea what the implications of this are. I'll try to continue digging into fork choice rules and some of the discussions around this specific spec update to gain additional context.

eserilev commented:
I can at least comment on the differences between the old vs new `calculate_committee_fraction` and why the results might be different
 
 ### The old `calculate_committee_fraction`
 
 we first calculate the average balance across all validators in a justified state. This is done by getting the total effective balance and dividing by the number of active validators.
 
 the next step is calculating the committee size, which is defined as the number of active validators divided by the slots per epoch.
 
 the committee weight is then calculated as the average balance multiplied by the committee size
 
 finally the committee fraction is calculated that the ( committee weight * proposer boost score ) / 100
 
 
 ### the new `calculate_committee_fraction`
 
 we define the committee weight as the total effective balance of all validators in a justified state divided by slots per epoch. this does not factor in the amount of validators in a justified state, which may play a role in the difference in calculations between the old vs new `calculate_committee_fraction`. 
 
testing locally, it seems the new committee weight calculation returns a slightly higher value than the old committee weight calculation

eserilev commented:
ok I thought about this some more and I think I understand whats going on here

in Rust the following calculations could return different results, even though they are mathematically equivalent

a * (b /c) vs  (a * b) / c

that's because w/ integer division, the remainder gets thrown away, so the order in which we divide and multiply plays a factor. 

so even though the old vs new committee_weight calculations are mathematically equivalent: 

(total_effective_balance / num_active_validators) * (num_active_validators / slots_per epoch) = total_effective_balance / slots_per_epoch

they dont return the exact same results due to the fact that the remainder gets thrown away during integer division

michaelsproul commented:
@eserilev Do you mind rebasing on/merging the latest `unstable`? That will fix the outstanding CI failures.

We'll try to review this ASAP and get it merged for 4.2.0

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into simplify-calculate-committee-fraction

eserilev commented:
@michaelsproul i merged the latest unstable. thank you!

michaelsproul reviewed (approved):
Looks good. Happy to merge.

michaelsproul commented:
bors r+
